File: 39_backtracking/39_8_white_hat_hacker.cpp
#include <iostream>
#include <string>
#include <vector>

const std::string CORRECT_PASSWORD = "bc";

bool check_password(const std::string& password) {
    return password == CORRECT_PASSWORD;
}

bool contains(const std::vector<char>& values, char target) {
    for (char value : values) {
        if (value == target) {
            return true;
        }
    }

    return false;
}

bool backtrack(
    int max_length,
    const std::vector<char>& elements,
    std::vector<char>& current,
    std::string& result) {
        if (static_cast<int>(current.size()) >= max_length) {
            return false;
        }

        for (char element : elements) {
            if (!contains(current, element)) {
                current.push_back(element);

                std::string candidate(current.begin(), current.end());
                if (check_password(candidate)) {
                    result = candidate;
                    return true;
                }

                if (backtrack(max_length, elements, current, result)) {
                    return true;
                }

                current.pop_back();
            }
        }

        return false;
    }

    std::string white_hat_hacker(int max_length) {
    std::vector<char> elements = {
        'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
        'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'};
    std::vector<char> current;
    std::string result;

    backtrack(max_length, elements, current, result);

    return result;
}

int main() {
    std::cout << white_hat_hacker(4) << std::endl;
    return 0;
}

File: 39_backtracking/39_8_white_hat_hacker.py
CORRECT_PASSWORD = "bc"
def check_password(s: str)->bool:
    return s == CORRECT_PASSWORD

def white_hat_hacker(max_length: int) -> str:
    elements = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
    res = ""

    def backtrack(cur: list):
        nonlocal res
        if len(cur) >= max_length:
            return

        for e in elements:
            if e not in cur:
                cur.append(e)
                print(cur)
                if check_password("".join(cur)):
                    res = "".join(cur)
                    return
                else:
                    backtrack(cur)

                cur.pop()
            
            if res != "":
                return

    backtrack([])
    return res

if __name__ == "__main__":
    print(white_hat_hacker(4))

